core/src/main/java/org/elasticsearch/transport/TransportChannelResponseHandler.java

```diff
@@ -67,15 +67,21 @@ protected TransportChannelResponseHandler(ESLogger logger, TransportChannel chan
 
     @Override
     public void handleResponse(T response) {
+        onFinished();
         try {
             channel.sendResponse(response);
         } catch (IOException e) {
             handleException(new TransportException(e));
         }
     }
 
+    protected void onFinished() {}
+
+    ;
+
     @Override
     public void handleException(TransportException exp) {
+        onFinished();
```

Comment:
I don't like adding onFinished method in this class. There is also no documentation for it. For example, this method can be invoked twice for same action if sending the response in handleResponse fails.